### PR TITLE
Bug fix in distributed metrics

### PR DESCRIPTION
**Pull request by AkshitaB**
Fixes https://github.com/allenai/allennlp/issues/4564
Remaining metrics

- [ ] Covariance

### CODE REVIEW

File: allennlp/training/metrics/covariance.py [hunk 1]
```diff
@@ -102,24 +101,24 @@ def __call__(
...
             device = gold_labels.device
-            _total_co_moment = torch.tensor(self._total_co_moment).to(device)
-            _total_count = torch.tensor(self._total_count).to(device)
-            _total_prediction_mean = torch.tensor(self._total_prediction_mean).to(device)
-            _total_label_mean = torch.tensor(self._total_prediction_mean).to(device)
-            dist.all_reduce(_total_co_moment, op=dist.ReduceOp.SUM)
+            delta_mean_prediction = torch.tensor(delta_mean_prediction).to(device)
+            delta_mean_label = torch.tensor(delta_mean_label).to(device)
+            delta_co_moment = torch.tensor(delta_co_moment).to(device)
+            _total_count = torch.tensor(updated_count).to(device)
```
Comment (epwalsh) [inline allennlp/training/metrics/covariance.py:111]:
```suggestion
            _total_count = torch.LongTensor(updated_count).to(device)
```

File: tests/training/metrics/attachment_scores_test.py [hunk 2]
```diff
@@ -148,3 +149,62 @@ def test_distributed_attachment_scores(self):
...
+    desired_values: Dict[str, Any],
+    exact: Union[bool, Tuple[float, float]] = True,
+):
+
+    kwargs = {}
+    # Use the arguments meant for the process with rank `global_rank`.
+    for argname in metric_kwargs:
+        kwargs[argname] = metric_kwargs[argname][global_rank]
+
+    for i in range(200):
```
Comment (epwalsh) [inline tests/training/metrics/attachment_scores_test.py:204]: Is this enough runs to cause the overflow error without the fixes you've made? Same question for all of the other tests.

Comment (AkshitaB) [reply tests/training/metrics/attachment_scores_test.py:204]: Yes, I confirmed that the test case got all nans and infs without the fixes I made.

**Comment by AkshitaB** [comment]
@epwalsh and I discussed and decided that distributed Covariance can be revisited later. Merging this now.

**Pull request merged**